src/journal-remote/journal-upload.c

```diff
@@ -234,12 +263,44 @@ int start_upload(Uploader *u,
                 easy_setopt(curl, CURLOPT_WRITEDATA, data,
                             LOG_ERR, return -EXFULL);
 
-                /* set where to read from */
-                easy_setopt(curl, CURLOPT_READFUNCTION, input_callback,
-                            LOG_ERR, return -EXFULL);
-
-                easy_setopt(curl, CURLOPT_READDATA, data,
-                            LOG_ERR, return -EXFULL);
+                if (arg_compress) {
+                        size_t step = 65535;
+                        void *buf = NULL;
+                        size_t size = step;
+                        size_t readSize = 0;
+                        while (u->uploading) {
+                                if (size - readSize < step) {
+                                        size = step + readSize;
+                                }
+                                if (buf == NULL) {
+                                        buf = malloc(size * sizeof(char));
+                                } else {
+                                        void *newBuf = realloc(buf, size * sizeof(char));
+                                        if (newBuf == NULL) {
+                                                log_error_errno(SYNTHETIC_ERRNO(EXFULL),
+                                                                "Failed to realloc buffer");
+                                                return -1;
+                                        }
+                                        buf = newBuf;
+                                }
+                                readSize += input_fn(buf, step, size - step, data);
+                        }
```

Comment:
so if i am getting this right you are reading the whole journal into memory here? sorry, but that's not really acceptable, tis might be ridiculously large. compression needs to happen progressively: i.e. compress, write out, compress more, write more, compress even more, write even more